server/src/main/java/org/apache/druid/segment/loading/SegmentLocalCacheManager.java

```diff
@@ -249,10 +279,14 @@ public List<DataSegment> getCachedSegments() throws IOException
          .emit();
     }
 
-    return cachedSegments;
+    return new ArrayList<>(cachedSegments);
```

Comment:
Return an immutable list:
```suggestion
    return List.copyOf(cachedSegments);
```